### ISSUE TITLE
Add support for DICOM dermoscopy objects

### ISSUE THREAD

**Issue by fedorov** [OHIF/Viewers#2934]
This follows up on the discussion with @Punzo on slack.

The samples can be accessed in https://challenge2020.isic-archive.com/.

If it turns out the effort required is significant, we will need to reconsider.

<img width="431" alt="image" src="https://user-images.githubusercontent.com/313942/190675715-d43abe86-d3e0-45ee-aa58-c014590a4d01.png">

FYI: @dclunie

**Comment by Punzo** [comment]
@fedorov it turns out that OHIFv2 can load dermoscopy data without any issue. 

I loaded locally the DICOM corrected Test Data (6,7 GB) from https://challenge2020.isic-archive.com/ . 

There is an issue: in the metadata the PhotometricInterpretation is set to YBR_FULL_422, however when downloading the image the length is rows * columns * 3 (instead * 2) (each pixel has 2 values, which can be converted to RGB/RGBA values, see [link](https://github.com/cornerstonejs/cornerstoneWADOImageLoader/blob/master/src/imageLoader/colorSpaceConverters/convertYBRFull422ByPixel.js)).
![Screenshot from 2022-09-20 12-46-16](https://user-images.githubusercontent.com/7985338/191238489-b6eb7714-f21f-4024-bf83-270f6521801d.png)

if I force the PhotometricInterpretation to RGB, the image is loaded properly

![Screenshot from 2022-09-20 12-47-17](https://user-images.githubusercontent.com/7985338/191238585-42f6c7da-61d0-4b41-9f0d-b249fb880bc0.png)

I could check if the image array is and in case convert thhe array as an RGB in https://github.com/cornerstonejs/cornerstoneWADOImageLoader/blob/master/src/imageLoader/convertColorSpace.js, but I would prefer not putting hacks in place for deffefctive datasets.


3DSlicer loads the images without problems, but I think it ignores the PhotometricInterpretation, and it just uses the SamplePerPixel attribute (which for YBR_FULL_422 is 3 by the standard, even if before the conversion to RGB/RGBA the array values for each pixel are 2).

Anyway, for datasets with proper metadata will be loaded with no problem with OHIFv2 with latests cornerstoneWADOImageLoader (see PR: https://github.com/OHIF/Viewers/pull/2944)

**Comment by dclunie** [comment]
@punzo, AFAIK, all these images are JPEG baseline compressed (Transfer Syntax 1.2.840.10008.1.2.4.50).

The DICOM Photometric Interpretation describes what is encoded in the compressed JPEG bitstream, not the RGB that might (or might not) come out of a JPEG decoder that converts the YCbCr chrominance downsampled compressed data into uncompressed RGB. See:

https://dicom.nema.org/medical/dicom/current/output/chtml/part05/sect_8.2.html#para_4bcb841e-c6bf-4e26-82a5-3fad3c942da0

and the notes after the table.

So, the logic in OHIF (or the image loader it is using) needs to account for whatever the JPEG decoder it is using actually does, and not be literally applied to the decompressed pixel data (i.e., it should know there are 3 pixels per sample in the decompressed data not 2 if chrominance downsampling has already been reversed), and that the color space has already been transformed by the decoder from YCbCr to RGB (both are usually the case for baseline JPEG decoders such as those that are libpeg-based).

Or to put it another way, each DECOMPRESSED pixel should have 3 values not 2 as you are assuming, regardless of the PhotometricInterpretation of the COMPRESSED pixel data being YBF_FULL_422. The only time that the PixelData length should be considered in this respect is when the encoding is "Native" (uncompressed), not "Encapsulated" (compressed), i.e., Note 2:

https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.6.3.html#para_442b66ec-6d91-42ba-bfeb-738c94d2d6d3

does not apply to these JPEG compressed images.

Your decoder should not be guessing that Photometric Interpretation might be "wrong", rather it should know exactly what the compressed pixel data decoder has done, and behave accordingly.

Be aware that not all JPEG compressed pixel data is YCbCr - sometimes it really was compressed as RGB components (e.g., Aperio WSI).

Also, for JPEG compressed images, though YBR_FULL_422 is always used when a YCbCr color transformation has been applied before compression,  the actual extent of chrominance downsampling varies in practice, and is not always 2x2:

https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.6.3.html#para_054ed567-c63e-4450-a128-0c8af7adc764

Further, some of the YBR family of Photometric Interpretations are used for non-JPEG compressed pixel data (e.g., some RLE or uncompressed ultrasound images, in which case a color space transformation +/- a chrominance channel upsampling may be required). That should not apply to these ISIC dermatology challenge images, though changes to your logic might affect handling of other images, which is why I mention it.

@chafey might want to comment on what Cornerstone does/should do in this respect for compressed images.

**Comment by Punzo** [comment]
Found the issue. The wado parser in cornerstoneWADOImageLoader sends a XMLHttpRequest  as type=application/octet-stream; transfer-syntax=* 
Then it uses the mime information from the server response to guess the transfer syntax of the donwloaded frame.

https://github.com/cornerstonejs/cornerstoneWADOImageLoader/blob/master/src/imageLoader/wadors/loadImage.js#L79
https://github.com/cornerstonejs/cornerstoneWADOImageLoader/blob/master/src/imageLoader/createImage.js#L238

However in the case of requests as type=application/octet-stream; transfer-syntax=* the right mime will not be known
(we had similar issues with SLIM: https://github.com/herrmannlab/dicom-microscopy-viewer/blob/master/src/pyramid.js#L426-L433). In fact in the case of this dermoscopy datasets, we would not get an a "image/jpeg" mime response (i.e. the frame is decompressed by the server and returned already as uncompressed color converted rgb image).

Therefore the cornerstoneWADOImageLoader will load internally the frame as if the Transfer Syntax is 1.2.840.10008.1.2 instead of 1.2.840.10008.1.2.4.50. This finallly generates issues, because color transformations are applied (not all decoders apply automatically the color transformations, so for a set of transfer syntaxes corresponding to such decoders, the color treansformations are appllied as in https://github.com/cornerstonejs/cornerstoneWADOImageLoader/blob/master/src/imageLoader/createImage.js#L238).

We could improve the cornerstoneWADOImageLoader either:
1) using the dicom-webclient instead to use a simple XMLHttpRequest (e.g., in Slim we use it to pass an array of transfer syntaxes https://github.com/herrmannlab/dicom-microscopy-viewer/blob/83739f1ba11840df7f4487cdf8015b7450ee419d/src/pyramid.js#L436-L491 and the first compatible one will used). In this way (having the "image/jpeg" mime of higher priority than application/octet-stream), we can get the original compressed jpeg file for the frame and we will decompress/apply color transformations client side.

or: 
2) After fetching the arrayBuffer, we could guess the right frame mime from the array buffer rather then from the mime information from the server response. E.g. similar to the approach that we use in SLIM: https://github.com/herrmannlab/dicom-microscopy-viewer/blob/83739f1ba11840df7f4487cdf8015b7450ee419d/src/webWorker/decodeAndTransformTask.js#L99-L150

@sedghi @swederik, it would be nice to have your feedback as well. Thanks!

**Comment by Punzo** [comment]
@fedorov moving this back in Progress

**Comment by Punzo** [comment]
open PR with fix at https://github.com/cornerstonejs/cornerstoneWADOImageLoader/pull/477

**Comment by fedorov** [comment]
Per @igoroctaviano, could it be related to https://github.com/OHIF/Viewers/issues/3354 ?

**Comment by jlulloaa** [comment]
> Per @igoroctaviano, could it be related to #3354 ?

Doesn't seem to be the same. It is not fixed with the update to cornerstone tools. From the data, it looks like the issue may be related to the way ohif viewer decode the RGB overlaid mask (I'll add more details about that directly in #3354). Also, in #3354 the thumbnails look ok, but not the images in the viewer, here instead, the images looked wrong in both, thumbnail and main viewer

**Comment by jlulloaa** [comment]
> A [latest update to ohif (v3-stable)](url) contains updates to cornerstone that solved the [issue with the RGB display](https://github.com/OHIF/Viewers/issues/2934#issuecomment-1252192617). Cornerstone's CPU-based thumbnails rendering didn't interpret correctly the RGB dermoscopy images, affecting how the thumbnails were displayed in OHIF. It was solved by changing the behaviour to GPU rendering by default ([cornerstonejs/cornerstone3D-beta#586](https://github.com/cornerstonejs/cornerstone3D-beta/pull/586)). This feature was transferred to OHIF in [PR 3372](https://github.com/OHIF/Viewers/pull/3372) and now available in [v3-stable](https://github.com/OHIF/Viewers/tree/v3-stable).
> 
> This was tested with a subset from "DICOM corrected Test Data (6,7 GB)", available at (https://challenge2020.isic-archive.com/) OHIF/Viewer (v3-stable, Cornerstone Tools v0.66.2): ![CPU rendering thumbnail](https://user-images.githubusercontent.com/84105116/237754676-ee3d2ead-ae59-44fa-af49-66239e079ab0.png)
> 
> OHIF/Viewer (v3-stable, Cornerstone Tools v0.67.1): ![GPU rendering thumbnail](https://user-images.githubusercontent.com/84105116/237748502-5f7b08b4-6fd7-446e-9732-582cea119e5b.png)
> 
> Same data loaded through [https://v3-demo.ohif.org](https://v3-demo.ohif.org/local?) ![v3-demo](https://user-images.githubusercontent.com/84105116/237810063-f031dec6-e2b2-4589-9cc6-fb0ec3670b7e.png)

As shown in this test, this issue is solved with the latest updates to v3-stable and the data is now displayed correctly (https://v3-demo.ohif.org/). 

I think this issue can be closed, @fedorov @igoroctaviano

**Comment by jlulloaa** [comment]
> Is it fixed in both v2 and v3, or only in v3?

Yes, I checked it works ok in both, v2 and v3

**Comment by fedorov** [comment]
Closing, per discussion today.

**Comment by fedorov** [comment]
@jlulloaa reopening - please confirm that this works with a Google DICOM Store!

**Comment by fedorov** [comment]
> please confirm that this works with a Google DICOM Store!

@igoroctaviano I guess @rodrigobasilio2022 will need to finish this up now?

**Comment by igoroctaviano** [comment]
@fedorov that's right, @rodrigobasilio2022 will continue this work.

**Comment by rodrigobasilio2022** [comment]
Due to my lack of permission, @igoroctaviano will test

**Comment by fedorov** [comment]
TODO for myself - add Rodrigo to 
<img width="1092" alt="image" src="https://github.com/OHIF/Viewers/assets/313942/331698f9-95fe-4b89-939d-b01816a3c7e7">

**Comment by fedorov** [comment]
@igoroctaviano the DICOM store in the above is not in a project that is in IDC, or which I can access.

I created a new DICOM store, imported all ISIC 2020 test images, and added @rodrigobasilio2022 and @igoroctaviano: `projects/idc-sandbox-000/locations/us-central1/datasets/isic/dicomStores/isic2020_test`

Please confirm you can access.

**Comment by igoroctaviano** [comment]
@fedorov looks like the dicom store does not contain the same data used here: https://github.com/OHIF/Viewers/issues/2934#issuecomment-1559695180

@rodrigobasilio2022 can you confirm that?

**Comment by rodrigobasilio2022** [comment]
This image is named ISIC_0520611.dcm

**Comment by fedorov** [comment]
The store used in https://github.com/OHIF/Viewers/issues/2934#issuecomment-1559695180 was set up by Jose, and I do not know anything about what it contained.

The store I set up should have all of the DICOM dermoscopy files included in the test dataset from https://challenge2020.isic-archive.com/.

**Comment by rodrigobasilio2022** [comment]
@igoroctaviano, you showed a bucket with few images, isn't it?

**Comment by fedorov** [comment]
I am confused. Are you looking for a bucket or for a DICOM store? Can't you test against the dicom store I set up for you?

**Comment by igoroctaviano** [comment]
@fedorov I tried (unsuccessful): http://localhost:3000/projects/idc-sandbox-000/locations/us-central1/datasets/isic/dicomStores/isic2020_test/study/1.3.6.1.4.1.5962.99.1.1195.2571.1620398270556.1.2.1

**Comment by fedorov** [comment]
I don't know if OAuth will work from localhost. Did that ever work?

**Comment by igoroctaviano** [comment]
> I don't know if OAuth will work from localhost. Did that ever work?

Yes, It works fine for other studies. localhost:3000 is a valid callback for this OIDC

**Comment by fedorov** [comment]
Permissions should be fine, I don't know what is going on then.

<img width="1452" alt="image" src="https://github.com/OHIF/Viewers/assets/313942/009dcf33-1063-424d-a19b-196f2e642706">

Here's the corresponding bucket: https://console.cloud.google.com/storage/browser/isic2020-test-sample.

**Comment by rodrigobasilio2022** [comment]
I don't have permission to list any item in this bucket

**Comment by fedorov** [comment]
This was confirmed to work against a GCP DICOM store.

**Comment by vicongit23** [comment]
Code cloned from https://github.com/OHIF/Viewers ( master branch) on July 20th still doesn't render RGB properly  
( independently of modality and transfer syntax). Does it mean that the fix just wasn't merged to master branch?

**Comment by rodrigobasilio2022** [comment]
Do you have an example image?

**Comment by vicongit23** [comment]
I do, cannot attach DICOM to comment though.

**Comment by fedorov** [comment]
There is an unresolved known issue related to v3 incorrectly handling multichannel images: https://github.com/OHIF/Viewers/issues/3354.

I guess in the context of this issue, the question is: why is it that dermoscopy images render fine, and other RGB examples in #3354 do not?

**Comment by vicongit23** [comment]
My examples are also US  ( as in 3354). Also PALETTE _COLOR doesn't work.

**Comment by vicongit23** [comment]
IMO dermoscopy images ( at least seen in OHIF v3 worklist) have photometric interpretation YBR_ICT

**Comment by rodrigobasilio2022** [comment]
I need to see the image to understand the problem. If you can upload to the dicom store ( I dont have the autority to let you), I can take a look in tje image.

**Comment by vicongit23** [comment]
Please look at 3354,
sample 3 displays the same presentation on similar US modality file

**Comment by rodrigobasilio2022** [comment]
You are from IDC? I don't have full access so I need the url to the image so I can access it

**Comment by fedorov** [comment]
No, I do not think @vicongit23 is from IDC. As suggested above, let's continue the discussion in #3354. Of note, there are links to the public samples from IDC in that issue.

**Comment by vicongit23** [comment]
> No, I do not think @vicongit23 is from IDC. As suggested above, let's continue the discussion in #3354. Of note, there are links to the public samples from IDC in that issue.

Correct

**Comment by fedorov** [comment]
This was broken https://d3w0jbfdn2j3pj.cloudfront.net/projects/idc-sandbox-000/locations/us-central1/datasets/isic/dicomStores/isic2020_test/study/1.3.6.1.4.1.5962.99.1.1195.6460.1620398329167.1.2.1

**Comment by igoroctaviano** [comment]
> This was broken https://d3w0jbfdn2j3pj.cloudfront.net/projects/idc-sandbox-000/locations/us-central1/datasets/isic/dicomStores/isic2020_test/study/1.3.6.1.4.1.5962.99.1.1195.6460.1620398329167.1.2.1

I opened this one today and it looked good, at least the first series. There's a specific broken series?

**Comment by igoroctaviano** [comment]
I tried again with master and I got different results this time, it's still broken:

![Screenshot 2023-10-20 at 15 28 42](https://github.com/OHIF/Viewers/assets/13886933/6ac289dd-a5e8-410b-8174-98bb766447d3)

**Comment by igoroctaviano** [comment]
> I tried again with master and I got different results this time, it's still broken:
> 
> ![Screenshot 2023-10-20 at 15 28 42](https://user-images.githubusercontent.com/13886933/277022704-6ac289dd-a5e8-410b-8174-98bb766447d3.png)

@sedghi FYI, looks like a regression

**Comment by igoroctaviano** [comment]
@sedghi do we have automation in place for the codec library? running snapshot testing and checking if the images change? I think this is very useful to avoid regression.

**Comment by sedghi** [comment]
sad :( 
No we don't have really but we should really. Can you give me that data in .dcm format?

**Comment by igoroctaviano** [comment]
> sad :( No we don't have really but we should really. Can you give me that data in .dcm format?

@fedorov is it ok to share this dataset?

**Comment by fedorov** [comment]
Yes! But note that it will be prudent for you to keep track of where a given test dataset is coming from, and make sure you are in compliance with the license. This dermoscopy sample is from ISIC: https://challenge2020.isic-archive.com/

**Comment by fedorov** [comment]
We do not seem to have any images in IDC that would be suitable to reproduce this issue.

<img width="777" alt="image" src="https://github.com/OHIF/Viewers/assets/313942/2d716b7e-499f-402e-8f6c-be6e0ae23d7c">
